File: last/memory_assistant_v1.0.0/background/llm_client.js
// LLM Client — 统一封装 LLM API 调用
// 目前使用 DeepSeek（与现有系统复用同一个 API Key）

const DEEPSEEK_ENDPOINT = "https://api.deepseek.com/v1/chat/completions";
const DEEPSEEK_MODEL = "deepseek-chat";

/**
 * 调用 LLM，返回解析后的 JSON 对象。
 * 支持三种容错解析：直接解析 → markdown 代码块 → 正则提取。
 *
 * @param {string} systemPrompt
 * @param {string} userPrompt
 * @param {string} apiKey
 * @returns {Promise<object>}  解析后的 JSON；若解析失败返回 {}
 */
export async function extractJson(systemPrompt, userPrompt, apiKey) {
  const text = await _call(systemPrompt, userPrompt, apiKey, 0.0, 2048);
  return _parseJson(text);
}

/**
 * 调用 LLM，返回纯文本。
 *
 * @param {string} systemPrompt
 * @param {string} userPrompt
 * @param {string} apiKey
 * @returns {Promise<string>}
 */
export async function summarize(systemPrompt, userPrompt, apiKey) {
  return _call(systemPrompt, userPrompt, apiKey, 0.3, 1024);
}

// ── 内部实现 ──────────────────────────────────────────────────────────────────

async function _call(systemPrompt, userPrompt, apiKey, temperature, maxTokens) {
  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), 60000); // 60s per request

  let resp;
  try {
    resp = await fetch(DEEPSEEK_ENDPOINT, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${apiKey}`,
      },
      body: JSON.stringify({
        model: DEEPSEEK_MODEL,
        temperature,
        max_tokens: maxTokens,
        messages: [
          { role: "system", content: systemPrompt },
          { role: "user",   content: userPrompt   },
        ],
      }),
      signal: controller.signal,
    });
  } finally {
    clearTimeout(timeoutId);
  }

  if (!resp.ok) {
    const err = await resp.text().catch(() => resp.statusText);
    throw new Error(`LLM API error ${resp.status}: ${err}`);
  }

  const data = await resp.json();
  return data.choices?.[0]?.message?.content ?? "";
}

function _parseJson(text) {
  // 1. 直接解析
  try { return JSON.parse(text); } catch { /* fall through */ }

  // 2. 提取 markdown 代码块中的 JSON
  const blockMatch = text.match(/```(?:json)?\s*([\s\S]*?)```/);
  if (blockMatch) {
    try { return JSON.parse(blockMatch[1].trim()); } catch { /* fall through */ }
  }

  // 3. 正则提取第一个 { ... } 或 [ ... ] 块
  const objMatch = text.match(/(\{[\s\S]*\}|\[[\s\S]*\])/);
  if (objMatch) {
    try { return JSON.parse(objMatch[1]); } catch { /* fall through */ }
  }

  console.warn("[llm_client] JSON parse failed, raw text:", text.slice(0, 200));
  return {};
}

File: _legacy/last/memory_assistant_v1.0.0/memory_transferor/src/memory_transferor/runtime/llm_client.py
from __future__ import annotations

import json
import os
import re
from typing import Any


class LLMClient:
    """Small OpenAI-compatible client used by the new memory path."""

    def __init__(
        self,
        *,
        api_key: str | None = None,
        base_url: str | None = None,
        model: str = "deepseek-chat",
    ) -> None:
        from openai import OpenAI

        self.client = OpenAI(
            api_key=api_key or os.environ.get("MWIKI_API_KEY") or os.environ.get("OPENAI_API_KEY"),
            base_url=base_url or os.environ.get("OPENAI_BASE_URL") or "https://api.deepseek.com/v1",
        )
        self.model = model

    def complete(self, system: str, user: str, *, temperature: float = 0.0) -> str:
        response = self.client.chat.completions.create(
            model=self.model,
            temperature=temperature,
            messages=[
                {"role": "system", "content": system},
                {"role": "user", "content": user},
            ],
        )
        return (response.choices[0].message.content or "").strip()

    def extract_json(self, system: str, user: str, *, temperature: float = 0.0) -> Any:
        text = self.complete(system, user, temperature=temperature)
        try:
            return json.loads(text)
        except json.JSONDecodeError:
            match = re.search(r"```(?:json)?\s*(.*?)```", text, re.DOTALL)
            if match:
                return json.loads(match.group(1))
            start = text.find("{")
            end = text.rfind("}")
            if start != -1 and end != -1 and end > start:
                return json.loads(text[start : end + 1])
            raise

